citest/gcp_testing/google_idp_session.py

```diff
@@ -0,0 +1,195 @@
+        .find('form')
+        .find_all('input'))
+    login_dict = {}
+    for input in login_form_inputs:
+      if input.has_attr('value'):
+        login_dict[input['name']] = input['value']
+
+    login_dict['Email'] = self.__login
+    login_dict['Passwd'] = self.__password
+    try:
+      post_response = self.__session.post(self.__user_approval_url,
+                                          data=login_dict)
+    except Exception as e:
+      __throw_http_session_exception(e, 'POST', self.__user_approval_url)
+
+    approval_form = BeautifulSoup(post_response.content).find('form')
+    approval_dict = {}
+    for input in approval_form.find_all('input'):
+      if input.has_attr('value'):
+        approval_dict[input['name']] = input['value']
+    approval_dict['submit_access'] = 'true'
+    try:
+      self.__session.post(approval_form['action'], data=approval_dict)
+    except Exception as e:
+      __throw_http_session_exception(e, 'POST', approval_form['action'])
+    self.__is_initialized = True
+
+
+  def terminate(self):
+    """Log out of the current authenticated Google session and the target
+    service's session if applicable.
+    """
+    if self.__logout_url:
+      try:
+        self.__session.post(self.__logout_url)
+      except Exception as e:
+        __throw_http_session_exception(e, 'POST', self.__logout_url)
+
+    try:
+      self.__session.post(GOOGLE_ACCOUNT_LOGOUT_URL)
+    except Exception as e:
+      __throw_http_session_exception(e, 'POST', GOOGLE_ACCOUNT_LOGOUT_URL)
+    self.__session.cookies = None
+
+
+  @property
+  def cookies(self):
+    """Get the HTTP session cookies from the current Google session.
+
+    Returns:
+      The collection of Cookie objects associated with the authenticated Google
+      session as a CookieJar object.
+      CookieJar documentation: https://docs.python.org/2.7/library/cookielib.html#cookiejar-and-filecookiejar-objects
+    """
+    if not self.__session.cookies:
+      raise AttributeError('No Http session cookies present!')
+    return self.__session.cookies
+
+
+  def mutate_request(self, req):
+    """Add the Google Session cookies to the request header.
+
+    Args:
+      req [urllib2.Request]: The request to add the Google Session cookies to.
+    """
+    if not isinstance(req, urllib2.Request):
+      raise TypeError('req not urllib2.Request: ' + req.__class__.__name__)
+    self.cookies.add_cookie_header(req)
+
+
+def __throw_http_session_exception(cause, method, url):
+  """Throw an exception caused by trying to access a url in the session flow.
+
+  Args:
+    cause [Exception]: Exception that was caught during the HTTP execution.
+    method [String]: Name of the HTTP method that was executed during the
+      failure.
+    url [String]: Target URL of the HTTP method.
+
+  Returns:
+    IDPSessionError with a pointer to the cause Exception and a reasonable
+      error message.
+  """
+  msg = "%s failed targeting %s" % (method, url)
```

Comment:
log this. I think. It shouldnt be noisy. In the context of citest, you want traceability to understand what happened or confirm that the tests were performed as you expected them to.